File: python/small_context_rag_solution/text.py
from __future__ import annotations

import math
import re
from collections import Counter

from .config import (
    CHUNK_SIZE,
    IMPORTANT_TERMS,
    STOPWORDS,
    SUMMARY_REDUCTION_INPUT_TOKEN_BUDGET,
)


def words(text: str) -> list[str]:
    return [
        word
        for word in re.findall(r"[a-z0-9]+", text.lower())
        if word not in STOPWORDS and len(word) > 1
    ]


def cosine_similarity(left: str, right: str) -> float:
    left_counts = Counter(words(left))
    right_counts = Counter(words(right))

    if not left_counts or not right_counts:
        return 0.0

    shared = set(left_counts) & set(right_counts)
    dot = sum(left_counts[word] * right_counts[word] for word in shared)
    left_norm = math.sqrt(sum(count * count for count in left_counts.values()))
    right_norm = math.sqrt(sum(count * count for count in right_counts.values()))
    return dot / (left_norm * right_norm)


def weighted_similarity(query: str, text: str) -> float:
    query_terms = set(words(query))
    text_terms = set(words(text))
    important_overlap = query_terms & text_terms & IMPORTANT_TERMS
    priority_overlap = important_overlap & {"budget", "fail", "failure", "failures"}
    return (
        cosine_similarity(query, text)
        + (0.15 * len(important_overlap))
        + (0.35 * len(priority_overlap))
    )


def approximate_tokens(text: str) -> int:
    return max(1, len(text) // 4)


def split_text(text: str) -> list[str]:
    chunks = []
    current_paragraphs = []
    current_size = 0

    for paragraph in re.split(r"\n\s*\n", text.strip()):
        paragraph = paragraph.strip()

        if not paragraph:
            continue

        if current_paragraphs and current_size + len(paragraph) > CHUNK_SIZE:
            chunks.append("\n\n".join(current_paragraphs))
            current_paragraphs = []
            current_size = 0

        current_paragraphs.append(paragraph)
        current_size += len(paragraph)

    if current_paragraphs:
        chunks.append("\n\n".join(current_paragraphs))

    return chunks


def summarize_text(text: str, max_sentences: int = 2) -> str:
    sentences = [
        sentence.strip()
        for sentence in re.split(r"(?<=[.!?])\s+", " ".join(text.split()))
        if sentence.strip()
    ]

    if len(sentences) <= max_sentences:
        return " ".join(sentences)

    scored_sentences = []

    for position, sentence in enumerate(sentences):
        sentence_words = words(sentence)
        term_score = sum(3 for word in sentence_words if word in IMPORTANT_TERMS)
        first_sentence_bonus = 1 if position == 0 else 0
        scored_sentences.append(
            (
                term_score + first_sentence_bonus,
                position,
                sentence,
            )
        )

    selected = sorted(
        scored_sentences,
        key=lambda item: (-item[0], item[1]),
    )[:max_sentences]
    selected.sort(key=lambda item: item[1])

    return " ".join(sentence for _score, _position, sentence in selected)


def pack_summaries_by_token_budget(
    summaries: list[str],
    token_budget: int,
) -> list[list[str]]:
    batches = []
    current_batch = []
    current_tokens = 0

    for summary in summaries:
        summary_tokens = approximate_tokens(summary)

        if current_batch and current_tokens + summary_tokens > token_budget:
            batches.append(current_batch)
            current_batch = []
            current_tokens = 0

        current_batch.append(summary)
        current_tokens += summary_tokens

    if current_batch:
        batches.append(current_batch)

    return batches


def force_summary_reduction_progress(summaries: list[str]) -> list[list[str]]:
    return [summaries[index : index + 2] for index in range(0, len(summaries), 2)]


def reduce_summary_batch(summaries: list[str]) -> str:
    combined = " ".join(summaries)
    return summarize_text(combined, max_sentences=3)


def recursively_reduce_summaries(summaries: list[str]) -> str:
    if not summaries:
        return "No summary available."

    current_summaries = summaries
    level = 1

    while len(current_summaries) > 1:
        batches = pack_summaries_by_token_budget(
            current_summaries,
            SUMMARY_REDUCTION_INPUT_TOKEN_BUDGET,
        )

        if len(batches) == len(current_summaries):
            batches = force_summary_reduction_progress(current_summaries)

        print(
            f"Reducing {len(current_summaries)} summaries into {len(batches)} "
            f"batch summaries at level {level}"
        )

        current_summaries = [reduce_summary_batch(batch) for batch in batches]
        level += 1

    return summarize_text(current_summaries[0], max_sentences=3)

File: typescript/src/text.ts
import {
  chunkSize,
  importantTerms,
  stopwords,
  summaryReductionInputTokenBudget,
} from "./config.js";

export function words(text: string): string[] {
  return Array.from(text.toLowerCase().matchAll(/[a-z0-9]+/g))
    .map((match) => match[0])
    .filter((word) => !stopwords.has(word) && word.length > 1);
}

export function cosineSimilarity(left: string, right: string): number {
  const leftCounts = countWords(words(left));
  const rightCounts = countWords(words(right));
  const shared = Array.from(leftCounts.keys()).filter((word) => rightCounts.has(word));

  if (shared.length === 0) {
    return 0;
  }

  const dot = shared.reduce(
    (sum, word) => sum + leftCounts.get(word)! * rightCounts.get(word)!,
    0,
  );
  const leftNorm = vectorNorm(leftCounts);
  const rightNorm = vectorNorm(rightCounts);
  return dot / (leftNorm * rightNorm);
}

export function weightedSimilarity(query: string, text: string): number {
  const queryTerms = new Set(words(query));
  const textTerms = new Set(words(text));
  const importantOverlap = Array.from(queryTerms).filter(
    (word) => textTerms.has(word) && importantTerms.has(word),
  );
  const priorityOverlap = importantOverlap.filter((word) =>
    ["budget", "fail", "failure", "failures"].includes(word),
  );

  return (
    cosineSimilarity(query, text)
    + 0.15 * importantOverlap.length
    + 0.35 * priorityOverlap.length
  );
}

function countWords(items: string[]): Map<string, number> {
  const counts = new Map<string, number>();

  for (const item of items) {
    counts.set(item, (counts.get(item) ?? 0) + 1);
  }

  return counts;
}

function vectorNorm(counts: Map<string, number>): number {
  return Math.sqrt(
    Array.from(counts.values()).reduce((sum, count) => sum + count * count, 0),
  );
}

export function approximateTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

export function splitText(text: string): string[] {
  const chunks: string[] = [];
  let currentParagraphs: string[] = [];
  let currentSize = 0;

  for (const paragraph of text.trim().split(/\n\s*\n/)) {
    const cleanParagraph = paragraph.trim();

    if (!cleanParagraph) {
      continue;
    }

    if (currentParagraphs.length > 0 && currentSize + cleanParagraph.length > chunkSize) {
      chunks.push(currentParagraphs.join("\n\n"));
      currentParagraphs = [];
      currentSize = 0;
    }

    currentParagraphs.push(cleanParagraph);
    currentSize += cleanParagraph.length;
  }

  if (currentParagraphs.length > 0) {
    chunks.push(currentParagraphs.join("\n\n"));
  }

  return chunks;
}

export function summarizeText(text: string, maxSentences = 2): string {
  const sentences = text
    .replace(/\s+/g, " ")
    .split(/(?<=[.!?])\s+/)
    .map((sentence) => sentence.trim())
    .filter(Boolean);

  if (sentences.length <= maxSentences) {
    return sentences.join(" ");
  }

  const scoredSentences = sentences.map((sentence, position) => {
    const termScore = words(sentence).filter((word) => importantTerms.has(word)).length * 3;
    const firstSentenceBonus = position === 0 ? 1 : 0;
    return {
      score: termScore + firstSentenceBonus,
      position,
      sentence,
    };
  });

  return scoredSentences
    .sort((left, right) => right.score - left.score || left.position - right.position)
    .slice(0, maxSentences)
    .sort((left, right) => left.position - right.position)
    .map((item) => item.sentence)
    .join(" ");
}

export function packSummariesByTokenBudget(
  summaries: string[],
  tokenBudget: number,
): string[][] {
  const batches: string[][] = [];
  let currentBatch: string[] = [];
  let currentTokens = 0;

  for (const summary of summaries) {
    const summaryTokens = approximateTokens(summary);

    if (currentBatch.length > 0 && currentTokens + summaryTokens > tokenBudget) {
      batches.push(currentBatch);
      currentBatch = [];
      currentTokens = 0;
    }

    currentBatch.push(summary);
    currentTokens += summaryTokens;
  }

  if (currentBatch.length > 0) {
    batches.push(currentBatch);
  }

  return batches;
}

export function forceSummaryReductionProgress(summaries: string[]): string[][] {
  const batches: string[][] = [];

  for (let index = 0; index < summaries.length; index += 2) {
    batches.push(summaries.slice(index, index + 2));
  }

  return batches;
}

export function reduceSummaryBatch(summaries: string[]): string {
  return summarizeText(summaries.join(" "), 3);
}

export function recursivelyReduceSummaries(summaries: string[]): string {
  if (summaries.length === 0) {
    return "No summary available.";
  }

  let currentSummaries = summaries;
  let level = 1;

  while (currentSummaries.length > 1) {
    let batches = packSummariesByTokenBudget(
      currentSummaries,
      summaryReductionInputTokenBudget,
    );

    if (batches.length === currentSummaries.length) {
      batches = forceSummaryReductionProgress(currentSummaries);
    }

    console.log(
      `Reducing ${currentSummaries.length} summaries into ${batches.length} batch summaries at level ${level}`,
    );

    currentSummaries = batches.map((batch) => reduceSummaryBatch(batch));
    level += 1;
  }

  return summarizeText(currentSummaries[0], 3);
}
